cmd/segment-reaper/observer.go

```diff
@@ -175,16 +175,38 @@ func (obsvr *observer) analyzeProject(ctx context.Context) error {
 			}
 
 			brokenObject := false
-			// expectedNumberOfSegments-1 because 'segments' doesn't contain last segment
 			if object.hasLastSegment {
+				includeLastSegment := true
+				segmentsCount := object.segments.Count()
+
+				// using 'expectedNumberOfSegments-1' because 'segments' doesn't contain last segment
 				switch {
+				case segmentsCount == 0:
+					// skip this case to avoid additional computation
 				case object.expectedNumberOfSegments == 0:
-					if !object.segments.IsSequence() {
-						brokenObject = true
+					// verify if initial sequence is valid
+					lastSequenceIndex := 0
+					for ; lastSequenceIndex < maxNumOfSegments; lastSequenceIndex++ {
+						has, err := object.segments.Has(lastSequenceIndex)
+						if err != nil {
+							panic(err)
+						}
+						if !has {
+							break
+						}
+					}
+					// unset valid sequence and mark as broken but don't remove last segment
+					for index := 0; index < lastSequenceIndex; index++ {
+						err := object.segments.Unset(index)
+						if err != nil {
+							panic(err)
+						}
 					}
-				case object.segments.Count() < int(object.expectedNumberOfSegments)-1:
 					brokenObject = true
```

Comment:
Is this case always a broken object?
For example, if the object has the bitmask in sequence until index 50 and after that all are set to 0 is it going this case to mark it as a borken object?

I thought that sometimes a valid object doesn't have the expected number of segments set.

Reply:
Value from `expectedNumberOfSegments` is set at the end of upload process and its a value that reflects how many segments were actually uploaded. If number of found segments is lower than this value then we are not able to download such object anymore.

Reply:
but this case is `case object.expectedNumberOfSegments == 0`

Reply:
ah sorry, on main page github attach this comment to a different line.

In case of `object.expectedNumberOfSegments == 0` we don't know how many segments we should expect in this object. We are iterating from first bit until we reach first `0`. Then we are unsetting valid sequence from the beginning (if exists) and if there are still segments after first `0` then we are marking what's left as a broken object but we are not removing `lastSegment`. We can compare `segmentsCount` with `lastSequenceIndex` to verify if there are still segments in this object after first `0` to mark/unmark object as broken if you think this will help with code understanding.

Reply:
Ok, yeah, that makes sense; it confuses to use `brokenObject` as a var name and I don't have any idea of a good name that fits for all the cases. We could do that check but also a comment should be fine too or perhaps better.

Reply:
comment added